solutions/usaco-596.mdx

```diff
@@ -97,3 +97,76 @@ void ff(int i, int j)
 </CPPSection>
 
 </LanguageSection>
+	
+## Alternate Faster Solution
+
+It is possible to make the following 2 observations:
+
+1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate. For example, if there are 2 regions, there must be 2 gates.
+2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous closed region if all of these conditions are met:
+	* FJ walks along an edge from point $a$ to $b$
+	* He did not walk from $a$ to $b$ in the past
+	* He did not walk from $b$ to $a$ in the past
+	* $b$ is a point he had already visited
+
+Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is about 10 times faster than flood filling due to $\log N$ lookup and insertion.
+	
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	freopen("gates.in", "r", stdin);
+	freopen("gates.out", "w", stdout);
+	
+	ios::sync_with_stdio(0);
+	cin.tie(0);
+
+	int n; cin >> n;
+	string s; cin >> s;
+	
+	set<pair<pair<int, int>, pair<int, int>>> visedge;
+	set<pair<int, int>> visnode;
+	int ans = 0;
+	pair<int, int> prev = {0, 0};
+	visnode.insert(prev);
+	
+	for (int i = 0; i < n; i++) {
+		int x = prev.first; int y = prev.second;
+		if (s[i] == 'N') {
+			x++;
+		} else if (s[i] == 'S') {
+			x--;
+		} else if (s[i] == 'E') {
+			y--;
+		} else {
+			y++;
+		}
+			  
+		if (visedge.find({{x, y}, prev}) == visedge.end() 
+				&& visnode.find({x, y}) != visnode.end()) {
+			ans++;
+		}
+			  
+		visedge.insert({{x, y}, prev});
+		visedge.insert({prev, {x, y}});
+        	visnode.insert({x, y});
+			  
+		prev.first = x;
+		prev.second = y;
+	}
+			  
```

Comment:
```suggestion
```

Reply:
Those lines are needed though. Either X or Y change.

Reply:
the suggestion is to delete the blank line (line 166) (not sure what you see but github diffs can be weird sometimes)

Reply:
what